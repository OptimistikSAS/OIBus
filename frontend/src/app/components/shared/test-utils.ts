import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
