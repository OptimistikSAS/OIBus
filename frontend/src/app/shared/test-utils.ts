import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
