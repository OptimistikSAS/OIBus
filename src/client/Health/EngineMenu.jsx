import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { FaEllipsisV } from 'react-icons/fa'
import { Dropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap'
import { useHistory } from 'react-router-dom'
import ConfirmationModal from '../components/ConfirmationModal.jsx'
import NewSouth from './NewSouth.jsx'
import NewNorth from './NewNorth.jsx'

const EngineMenu = ({ onRestart, onShutdown }) => {
  const [southModal, setSouthModal] = useState(false)
  const [northModal, setNorthModal] = useState(false)
  const [restartShow, setRestartShow] = useState(false)
  const [shutdownShow, setShutdownShow] = useState(false)
  const [dropdownOpen, setDropdownOpen] = useState(false)
  const toggle = () => setDropdownOpen((prevState) => !prevState)

  const history = useHistory()

  const handleGoToConnector = (pathname) => {
    history.push({ pathname })
  }
  const titleRestart = 'Restart'
  const titleShutDown = 'Shutdown'
  const bodyRestart = 'Confirm restart?'
  const bodyShutDown = 'Confirm shutdown?'

  return (
    <>

      <Dropdown
        isOpen={dropdownOpen}
        toggle={toggle}
        direction="down"
      >
        <DropdownToggle size="sm" className="icon-dropdown">
          <FaEllipsisV id="dropdown-toggle" className="icon-dropdown-ellipsis" />
        </DropdownToggle>

        <DropdownMenu style={{ textAlign: 'center' }}>
          <DropdownItem id="add-north" className="dropdown-item" onClick={() => setNorthModal(true)}>
            Add North
          </DropdownItem>

          <DropdownItem id="add-south" className="dropdown-item" onClick={() => setSouthModal(true)}>
            Add South
          </DropdownItem>

          <DropdownItem className="dropdown-item">
            <div
              id="oi-settings"
              aria-hidden="true"
              role="button"
              onClick={() => {
                handleGoToConnector('/Engine/')
              }}
            >
              Settings
            </div>
          </DropdownItem>

          <DropdownItem id="restart" className="dropdown-item" onClick={() => setRestartShow(true)}>
            Restart
          </DropdownItem>
          <ConfirmationModal
            title={titleRestart}
            body={bodyRestart}
            onConfirm={onRestart}
            isOpen={restartShow}
            toggle={() => setRestartShow(false)}
          />

          <DropdownItem id="shutdown" className="dropdown-item" onClick={() => setShutdownShow(true)}>
            Shutdown
          </DropdownItem>
          <ConfirmationModal
            title={titleShutDown}
            body={bodyShutDown}
            onConfirm={onShutdown}
            isOpen={shutdownShow}
            toggle={() => setShutdownShow(false)}
          />
        </DropdownMenu>
      </Dropdown>
      <NewSouth modal={southModal} toggle={() => setSouthModal(false)} />
      <NewNorth modal={northModal} toggle={() => setNorthModal(false)} />

    </>
  )
}

EngineMenu.propTypes = {
  onRestart: PropTypes.func.isRequired,
  onShutdown: PropTypes.func.isRequired,
}

export default EngineMenu
