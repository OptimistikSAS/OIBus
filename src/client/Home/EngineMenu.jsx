import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { FaEllipsisV } from 'react-icons/fa'
import { Dropdown, DropdownToggle, DropdownMenu, DropdownItem } from 'reactstrap'
import { useNavigate } from 'react-router-dom'
import ConfirmationModal from '../components/ConfirmationModal.jsx'
import NewSouth from './NewSouth.jsx'
import NewNorth from './NewNorth.jsx'

const EngineMenu = ({ onRestart, onShutdown }) => {
  const [southModal, setSouthModal] = useState(false)
  const [northModal, setNorthModal] = useState(false)
  const [restartShow, setRestartShow] = useState(false)
  const [shutdownShow, setShutdownShow] = useState(false)
  const [dropdownOpen, setDropdownOpen] = useState(false)
  const toggle = () => setDropdownOpen((prevState) => !prevState)
  const navigate = useNavigate()

  return (
    <>
      <Dropdown
        isOpen={dropdownOpen}
        toggle={toggle}
        direction="down"
      >
        <DropdownToggle size="sm" id="dropdown-toggle" className="p-0 m-0 oi-dropdown-toggle">
          <FaEllipsisV />
        </DropdownToggle>
        <DropdownMenu>
          <DropdownItem id="add-north" onClick={() => setNorthModal(true)}>
            Add North
          </DropdownItem>
          <DropdownItem id="add-south" onClick={() => setSouthModal(true)}>
            Add South
          </DropdownItem>
          <DropdownItem
            id="oi-settings"
            onClick={() => {
              navigate('/engine/')
            }}
          >
            Settings
          </DropdownItem>
          <DropdownItem id="restart" onClick={() => setRestartShow(true)}>
            Restart
          </DropdownItem>
          <ConfirmationModal
            title="Restart"
            body="Confirm restart?"
            onConfirm={onRestart}
            isOpen={restartShow}
            toggle={() => setRestartShow(false)}
          />
          <DropdownItem id="shutdown" onClick={() => setShutdownShow(true)}>
            Shutdown
          </DropdownItem>
          <ConfirmationModal
            title="Shutdown"
            body="Confirm shutdown?"
            onConfirm={onShutdown}
            isOpen={shutdownShow}
            toggle={() => setShutdownShow(false)}
          />
          <DropdownItem
            id="oi-about"
            onClick={() => {
              navigate('/about/')
            }}
          >
            About
          </DropdownItem>
        </DropdownMenu>

      </Dropdown>
      <NewSouth modal={southModal} toggle={() => setSouthModal(false)} />
      <NewNorth modal={northModal} toggle={() => setNorthModal(false)} />
    </>
  )
}

EngineMenu.propTypes = {
  onRestart: PropTypes.func.isRequired,
  onShutdown: PropTypes.func.isRequired,
}

export default EngineMenu
