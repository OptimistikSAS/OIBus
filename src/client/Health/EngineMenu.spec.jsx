/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act, Simulate } from 'react-dom/test-utils'

import EngineMenu from './EngineMenu.jsx'

import activeConfig from '../../../tests/testConfig'

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => (
  { useHistory: () => ({ push: mockHistoryPush }) }
))

const restartFunction = jest.fn()
const shutdownFunction = jest.fn()

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
  jest.clearAllMocks()
})

React.useContext = jest.fn().mockReturnValue({ activeConfig })
describe('EngineMenu', () => {
  test('display EngineMenu page based on config', async () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })

  test('check restart engine', () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    Simulate.click(document.getElementById('dropdown-toggle'))
    Simulate.click(document.getElementById('icon-restart'))
    Simulate.click(document.getElementById('icon-confirm'))
    expect(restartFunction).toBeCalled()
    expect(container).toMatchSnapshot()
  })

  test('check cancel restart', () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    Simulate.click(document.getElementById('dropdown-toggle'))
    Simulate.click(document.getElementById('icon-restart'))
    Simulate.click(document.getElementById('cancel-button'))
    expect(restartFunction).not.toBeCalled()
    expect(container).toMatchSnapshot()
  })

  test('check shutdown engine', () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    Simulate.click(document.getElementById('dropdown-toggle'))
    Simulate.click(document.getElementById('icon-shutdown'))
    Simulate.click(document.getElementById('icon-confirm'))
    // expect(shutdownFunction).toBeCalled()
    expect(container).toMatchSnapshot()
  })

  test('check cancel shutdown', () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    Simulate.click(document.getElementById('dropdown-toggle'))
    Simulate.click(document.getElementById('icon-shutdown'))
    Simulate.click(document.getElementById('cancel-button'))
    // expect(shutdownFunction).not.toBeCalled()
    expect(container).toMatchSnapshot()
  })

  test('go to engine settings', () => {
    act(() => {
      ReactDOM.render(
        <EngineMenu onRestart={restartFunction} onShutdown={shutdownFunction} />,
        container,
      )
    })
    Simulate.click(document.getElementById('dropdown-toggle'))
    Simulate.click(document.getElementById('icon-settings'))
    expect(mockHistoryPush).toBeCalledWith({ pathname: '/Engine/' })
    expect(container).toMatchSnapshot()
  })
})
