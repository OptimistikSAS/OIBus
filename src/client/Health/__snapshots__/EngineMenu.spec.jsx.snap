// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngineMenu check cancel restart 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;

exports[`EngineMenu check cancel shutdown 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;

exports[`EngineMenu check restart engine 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;

exports[`EngineMenu check shutdown engine 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;

exports[`EngineMenu display EngineMenu page based on config 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;

exports[`EngineMenu go to engine settings 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="icon-dropdown btn btn-secondary btn-sm"
      type="button"
    >
      <svg
        class="icon-dropdown-ellipsis"
        fill="currentColor"
        height="1em"
        id="dropdown-toggle"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="dropdown-menu"
      role="menu"
      style="text-align: center;"
      tabindex="-1"
    >
      <button
        class="dropdown-item dropdown-item"
        id="add-north"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add North
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="add-south"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Add South
      </button>
      <button
        class="dropdown-item dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          aria-hidden="true"
          id="oi-settings"
          role="button"
        >
          Settings
        </div>
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="restart"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Restart
      </button>
      <button
        class="dropdown-item dropdown-item"
        id="shutdown"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        Shutdown
      </button>
    </div>
  </div>
</div>
`;
